src/aind_behavior_curriculum/tmp.py

```diff
@@ -0,0 +1,106 @@
+"""
+Work in Progress Curriculum
+"""
+
+from collections import defaultdict
+from typing import Any
+
+class Task:
+    """
+    Defined in exisiting PR
+    """
+    pass
+
+class Metric:
+    """
+    Initalizes a connection to relevant database (referencing some configuration file)
+    User extends this with their own metric implementation based on data inside the database.
+    """
+    pass
+
+
+class Evaluation:
+    """
+    Serializable Callable that performs evaluation,
+    based on set of input metrics. Also serializes metrics.
+    """
+    pass
+
+
+# TODO: De/serialization
+# Define de/serialization such that calling serialization on Curriculum
+# recursively serializes graph components: (Task, Metric, Evaluation)
+class Curriculum:
+    def __init__(self) -> None:
+        TaskId = int
+        # ID tasks on Task.name, a required field.
+        # Tasks are allowed to change parameters.
+        self.tasks: dict[str, TaskId] = {}
+        # Each node references a collection of (edge, destinatation node) associations.
+        # The collection is ordered by order of Curriculum API calls.
+        self.graph: dict[TaskId, dict[TaskId, Evaluation]] = defaultdict(dict)
+
+    def add_node(self, task: Task):
+        self.tasks[task.name] = len(self.tasks)
+
+    def add_connection(self, start_task: str, next_task: str, eval: Evaluation):
+        task_id_1 = self.tasks[start_task]
+        task_id_2 = self.tasks[next_task]
+        self.graph[task_id_1][task_id_2] = eval
+
+    def export_visual(self, ...):
+        # Export a visual representation of the curriculum to inspect if it is correct.
+        # (See Han's Code)
+        pass
+
+
+def combine_curriculums(curr_1: Curriculum, curr_2: Curriculum) -> Curriculum:
```

Comment:
Combining curriculums is a useful feature. I think we will need to provide more precise inputs to combine curriculums. I can imagine it working something like "Connect curriculum A, node/stage X with curriculum B at node/stage Y"

Reply:
Not sure you need this as a function to combine the two curricula. You can always merge 2 curricula even if there are no transitions between them. At least they look fully composable to me. Sure you will have to define rules if you want to have a final curriculum that makes sense, but from the computational point of view i am not sure why it would be a problem. The exception here is what to do if the two curricula have the same task/stage key in the dictionaries. I can see two things happening:
1. Throw if the union of the keys is not unique
2. Find equal keys, and take the union of the transition rules
This could even be an optional parameter in the concat function (e.g. force_outter_join = True)